apps/design/backend/migrations/1742422332457_election-content-tables.js

```diff
@@ -0,0 +1,431 @@
+/* eslint-disable @typescript-eslint/no-var-requires */
+const { throwIllegalValue } = require('@votingworks/basics');
+const { hasSplits } = require('@votingworks/types');
+const { PgLiteral } = require('node-pg-migrate');
+
+/**
+ * @type {import('node-pg-migrate').ColumnDefinitions | undefined}
+ */
+exports.shorthands = {
+  id: { type: 'text', primaryKey: true },
+  election_id: {
+    type: 'text',
+    notNull: true,
+    references: 'elections',
+    onDelete: 'CASCADE',
+  },
+  created_at: {
+    type: 'timestamptz',
+    notNull: true,
+    default: new PgLiteral('current_timestamp'),
+  },
+};
+
+/**
+ * Migrate election JSON data into normalized tables. Steps:
+ * 1. Add tables for each entity within an election/add columns to the elections
+ *    table for election attributes
+ * 2. Copy data from the election JSON into the new tables
+ * 3. Drop the original JSON columns from the elections table
+ *
+ * @param pgm {import('node-pg-migrate').MigrationBuilder}
+ * @param run {() => void | undefined}
+ * @returns {Promise<void>}
+ */
+exports.up = async (pgm) => {
+  /**
+   * Step 1: Add new columns/tables
+   */
+  pgm.addColumns('elections', {
+    type: { type: 'text' },
+    title: { type: 'text' },
+    date: { type: 'date' },
+    jurisdiction: { type: 'text' },
+    state: { type: 'text' },
+    seal: { type: 'text' },
+    ballot_paper_size: { type: 'text' },
+  });
+  pgm.createTable(
+    'districts',
+    {
+      id: 'id',
+      election_id: 'election_id',
+      name: { type: 'text', notNull: true },
+      created_at: 'created_at',
+    },
+    { ifNotExists: true }
+  );
+  pgm.createTable(
+    'precincts',
+    {
+      id: 'id',
+      election_id: 'election_id',
+      name: { type: 'text', notNull: true },
+      created_at: 'created_at',
+    },
+    { ifNotExists: true }
+  );
+  pgm.createTable(
+    'precinct_splits',
+    {
+      id: 'id',
+      precinct_id: {
+        type: 'text',
+        notNull: true,
+        references: 'precincts',
+        onDelete: 'CASCADE',
+      },
+      name: { type: 'text', notNull: true },
+      nh_options: { type: 'jsonb', notNull: true },
+      created_at: 'created_at',
+    },
+    { ifNotExists: true }
+  );
+  pgm.createTable(
+    'districts_precincts',
```

Comment:
Do we need uniqueness guarantees here? Primary key on `[district_id, precinct_id]`?
(Similar question for the other associative tables)

Reply:
Yes, good catch, thanks